skyportal/handlers/api/allocation.py

```diff
@@ -58,50 +58,61 @@ def get(self, allocation_id=None):
-                instruments_subquery = sa.select(Instrument.id).filter(
-                    Instrument.api_classname_obsplan.isnot(None)
+        with self.Session() as session:
+            allocations = Allocation.select(self.current_user)
+
+            if allocation_id is not None:
+                try:
+                    allocation_id = int(allocation_id)
+                except ValueError:
+                    return self.error("Allocation ID must be an integer.")
+                allocations = allocations.where(Allocation.id == allocation_id)
+                allocation = session.scalars(allocations).first()
+                if allocation is None:
+                    return self.error("Could not retrieve allocation.")
+                return self.success(data=allocation)
+
+            instrument_id = self.get_query_argument('instrument_id', None)
+            if instrument_id is not None:
+                allocations = allocations.where(
+                    Allocation.instrument_id == instrument_id
                 )
 
-                allocations = allocations.join(
-                    instruments_subquery,
-                    Allocation.instrument_id == instruments_subquery.c.id,
-                )
-            else:
-                return self.error(
-                    f"apitype can only be api_classname or api_classname_obsplan, not {apitype}"
-                )
+            apitype = self.get_query_argument('apiType', None)
+            if apitype is not None:
+                if apitype == "api_classname":
+                    instruments_subquery = (
+                        sa.select(Instrument.id)
+                        .filter(Instrument.api_classname.isnot(None))
+                        .distinct()
+                        .subquery()
+                    )
+
+                    allocations = allocations.join(
+                        instruments_subquery,
+                        Allocation.instrument_id == instruments_subquery.c.id,
+                    )
+                elif apitype == "api_classname_obsplan":
+                    instruments_subquery = (
+                        sa.select(Instrument.id)
+                        .filter(Instrument.api_classname_obsplan.isnot(None))
+                        .distinct()
+                        .subquery()
+                    )
+
+                    allocations = allocations.join(
+                        instruments_subquery,
+                        Allocation.instrument_id == instruments_subquery.c.id,
+                    )
+                else:
+                    return self.error(
+                        f"apitype can only be api_classname or api_classname_obsplan, not {apitype}"
+                    )
 
-        allocations = allocations.all()
-        self.verify_and_commit()
-        return self.success(data=allocations)
+            allocations = session.scalars(allocations).all()
+            self.verify_and_commit()
```

Comment:
```suggestion
            session.commit()
```

Reply:
Are you sure we even need a `commit` here?

Reply:
Good point!